datastore/sqlite.nim

```diff
@@ -0,0 +1,238 @@
+import pkg/questionable
+import pkg/questionable/results
+import pkg/sqlite3_abi
+import pkg/upraises
+
+push: {.upraises: [].}
+
+# Adapted from:
+# https://github.com/status-im/nwaku/blob/master/waku/v2/node/storage/sqlite.nim
+
+type
+  AutoDisposed*[T: ptr|ref] = object
+    val*: T
+
+  DataProc* = proc(s: RawStmtPtr) {.closure.}
+
+  NoParams* = tuple # empty tuple
+
+  NoParamsStmt* = SQLiteStmt[NoParams, void]
+
+  RawStmtPtr* = ptr sqlite3_stmt
+
+  SQLite* = ptr sqlite3
+
+  SQLiteStmt*[Params, Res] = distinct RawStmtPtr
+
+proc sqlite3_bind_text(
+  pstmt: ptr sqlite3_stmt,
+  param: cint,
+  value: pointer,
+  n: cint,
+  dispose: proc (v: pointer) {.cdecl.}): cint {.importc, cdecl.}
+
+proc bindParam(
+  s: RawStmtPtr,
+  n: int,
+  val: auto): cint =
+
+  when val is openarray[byte]|seq[byte]:
+    if val.len > 0:
+      try:
+        # `SQLITE_TRANSIENT` "indicate[s] that the object is to be copied prior
+        # to the return from sqlite3_bind_*(). The object and pointer to it
+        # must remain valid until then. SQLite will then manage the lifetime of
+        # its private copy."
+        sqlite3_bind_blob(s, n.cint, unsafeAddr val[0], val.len.cint,
+          SQLITE_TRANSIENT)
+      except Exception as e:
+        raise (ref Defect)(msg: $e.msg)
+    else:
+      sqlite3_bind_blob(s, n.cint, nil, 0.cint, nil)
+  elif val is int32:
+    sqlite3_bind_int(s, n.cint, val)
+  elif val is uint32:
+    sqlite3_bind_int(s, int(n).cint, int(val).cint)
```

Comment:
yep, correct it should use the 64 bit version.

Reply:
This was copy-pasted from [nwaku's sqlite.nim](https://github.com/status-im/nwaku/blob/master/waku/v2/node/storage/sqlite.nim#L130-L148). That doesn't make it wrong or right; however, to my understanding:
* `int(n)` will convert the unsigned `uint32` to Nim's [default integer type](https://nim-lang.org/docs/system.html#int)
* `int(n).cint` thus converts from `uint32` to [*the type `int` in C*](https://nim-lang.org/docs/system.html#cint)

Reply:
I suspect that issues might arise in the way that sqlite represents this values internally. If for some reason sqlite decides to pack values ignoring the sign bit, things can get messy. 

I've found at least one example where uint is stored using the 64 bit version - https://cpp.hotexamples.com/site/file?hash=0xd834bc2db84f0c7ee36fd30ecc9c3625e7c0bcd6d32f8fb48ed8c49d5b05a012&fullName=QGIS-master/src/providers/spatialite/qsql_spatialite.cpp&project=Br1ndavoine/QGIS.

Reply:
No problem, I made pull request there too :) https://github.com/status-im/nwaku/pull/1014

1. n is already int, don't mix it with val !
2. when val >= 1<<31, using sqlite3_bind_int should store it as negative value. C API accepts signed value, it has no way to know that you had an unsigned one. So, for values >= 1<<31, you have to pass them as 64-bit ones (ideally, we should also make a test passing `uint32(1) << 31` and then requesting it as 64-bit value and checking that it's indeed `1u64 << 31`)

Although, honestly speaking, we don't use uint32 values in nimstore-sqlite backend. They, may be, too :)

Reply:
I may misunderstand something here and am always happy to learn and be better informed. 😄 

The opening context is, per the commit I pushed recently, [L32](https://github.com/status-im/nim-datastore/blob/initial_impl/datastore/sqlite.nim#L32): `when val is openarray[byte]|seq[byte]`. To my understanding of the Nim manual, [when](https://nim-lang.org/docs/manual.html#statements-and-expressions-when-statement) is evaluated at compile time.

So, when we consider [L44-L45](https://github.com/status-im/nim-datastore/blob/initial_impl/datastore/sqlite.nim#L44-L45) (again, line numbers are per my latest commit), there's generated code that's guaranteed to be working with a `val` that's a `uint32`.

Therefore, per `int(n).cint` it seems we'd always be passing to SQLite the largest signed integer type supported by the compiler.

SQLite's docs say:
> The INTEGER storage class, for example, includes 7 different integer datatypes of different lengths. [This makes a difference on disk.](https://www.sqlite.org/fileformat2.html#record_format) But as soon as INTEGER values are read off of disk and into memory for processing, they are converted to the most general datatype (8-byte signed integer). And so for the most part, "storage class" is indistinguishable from "datatype" and the two terms can be used interchangeably.

Am I missing anything?

Reply:
>  But `then int(val).cint` will convert it into 32-bit integer

According to my understanding of the Nim manual, per compile-time `when ...: elif val is uint32:`, we're already dealing with a 32-bit unsigned integer as the input at runtime, never any other type, so I don't understand how information could be lost.

Thanks for helping me to understand it better 😅

Reply:
`n` represents the zero-indexed column number so practically it's fine to do `n.cint`. If we want to be more precise we could specify `n: Natural` or a range that tops out at `hight(int32)`. However, neither `template bindParams` nor `proc bindParam` are exported, and the way they're currently used (with index numbers derived from tuples), I think it's unlikely we'll encounter a problem.

I'm not sure why the nwaku folks hadn't updated `int(n).cint` in the `elif val is uint32` branch of the `when` statement, but I already changed it yesterday since there's no reason it can't be `n.cint` like the others, i.e. unless we want/need to be more precise. It was probably just overlooked at some point.

What we really care about is `val`.

Consider this:
```nim
echo low(uint32)
echo high(uint32)
echo cast[int32](low(uint32))
echo cast[int32](high(uint32))
echo cast[uint32](cast[int32](low(uint32)))
echo cast[uint32](cast[int32](high(uint32)))

echo ""

echo low(int32)
echo high(int32)
echo cast[uint32](low(int32))
echo cast[uint32](high(int32))
echo cast[int32](cast[uint32](low(int32)))
echo cast[int32](cast[uint32](high(int32)))
```
prints:
```
0
4294967295
0
-1
0
4294967295

-2147483648
2147483647
2147483648
2147483647
-2147483648
2147483647
```

So as you say, *"since we're explicitly matching for a `uint32` here with the when statement, it should be safe to convert the `uint32` to an `int32`, since they are both 4 bytes wide."* That's exactly what I had in mind yesterday.

I'm happy to make the suggested change from `sqlite3_bind_int(s, n.cint, int(val).cint)` to `sqlite3_bind_int64(s, n.cint, val)` if it's for the best. I just wanted to understand under what conditions we might run into a problem without the change, and I'm still not convinced we would.